diff --git a/TMessagesProj/src/main/java/org/telegram/ui/Cells/DrawerUserCell.java b/TMessagesProj/src/main/java/org/telegram/ui/Cells/DrawerUserCell.java
index 1557a7847..f08e63e02 100644
--- a/TMessagesProj/src/main/java/org/telegram/ui/Cells/DrawerUserCell.java
+++ b/TMessagesProj/src/main/java/org/telegram/ui/Cells/DrawerUserCell.java
@@ -10,13 +10,14 @@ package org.telegram.ui.Cells;
 
 import android.content.Context;
 import android.graphics.Canvas;
-import android.graphics.Path;
 import android.graphics.RectF;
 import android.graphics.drawable.Drawable;
 import android.view.Gravity;
 import android.view.accessibility.AccessibilityNodeInfo;
 import android.widget.FrameLayout;
 
+import com.evildayz.code.telegraher.ThePenisMightierThanTheSword;
+import org.telegram.PhoneFormat.PhoneFormat;
 import org.telegram.messenger.AndroidUtilities;
 import org.telegram.messenger.ContactsController;
 import org.telegram.messenger.Emoji;
@@ -24,6 +25,7 @@ import org.telegram.messenger.MessagesController;
 import org.telegram.messenger.MessagesStorage;
 import org.telegram.messenger.NotificationCenter;
 import org.telegram.messenger.NotificationsController;
+import org.telegram.messenger.SharedConfig;
 import org.telegram.messenger.UserConfig;
 import org.telegram.tgnet.TLRPC;
 import org.telegram.ui.ActionBar.SimpleTextView;
@@ -38,6 +40,7 @@ import org.telegram.ui.Components.Premium.PremiumGradient;
 public class DrawerUserCell extends FrameLayout implements NotificationCenter.NotificationCenterDelegate {
 
     private SimpleTextView textView;
+    private SimpleTextView phoneTextView;
     private BackupImageView imageView;
     private AvatarDrawable avatarDrawable;
     private GroupCreateCheckBox checkBox;
@@ -69,6 +72,15 @@ public class DrawerUserCell extends FrameLayout implements NotificationCenter.No
         status = new AnimatedEmojiDrawable.SwapAnimatedEmojiDrawable(textView, AndroidUtilities.dp(20));
         textView.setRightDrawable(status);
 
+        phoneTextView = new SimpleTextView(context);
+        phoneTextView.setTextColor(Theme.getColor(Theme.key_chats_menuItemText));
+        phoneTextView.setTextSize(15);
+        phoneTextView.setMaxLines(1);
+        phoneTextView.setGravity(Gravity.LEFT);
+        if (!MessagesController.getGlobalTelegraherSettings().getBoolean("HidePhoneNumberOnLeftPanel", false))
+            addView(phoneTextView, LayoutHelper.createFrame(LayoutHelper.MATCH_PARENT, LayoutHelper.WRAP_CONTENT, Gravity.LEFT | Gravity.BOTTOM, 72, 0, 60, 0));
+
+
         checkBox = new GroupCreateCheckBox(context);
         checkBox.setChecked(true, false);
         checkBox.setCheckScale(0.9f);
@@ -88,7 +100,7 @@ public class DrawerUserCell extends FrameLayout implements NotificationCenter.No
     protected void onAttachedToWindow() {
         super.onAttachedToWindow();
         textView.setTextColor(Theme.getColor(Theme.key_chats_menuItemText));
-        for (int i = 0; i < UserConfig.MAX_ACCOUNT_COUNT; i++){
+        for (int i : SharedConfig.activeAccounts) {
             NotificationCenter.getInstance(i).addObserver(this, NotificationCenter.currentUserPremiumStatusChanged);
             NotificationCenter.getInstance(i).addObserver(this, NotificationCenter.updateInterfaces);
         }
@@ -98,7 +110,7 @@ public class DrawerUserCell extends FrameLayout implements NotificationCenter.No
     @Override
     protected void onDetachedFromWindow() {
         super.onDetachedFromWindow();
-        for (int i = 0; i < UserConfig.MAX_ACCOUNT_COUNT; i++){
+        for (int i : SharedConfig.activeAccounts)  {
             NotificationCenter.getInstance(i).removeObserver(this, NotificationCenter.currentUserPremiumStatusChanged);
             NotificationCenter.getInstance(i).removeObserver(this, NotificationCenter.updateInterfaces);
         }
@@ -139,6 +151,7 @@ public class DrawerUserCell extends FrameLayout implements NotificationCenter.No
             text = Emoji.replaceEmoji(text, textView.getPaint().getFontMetricsInt(), AndroidUtilities.dp(20), false);
         } catch (Exception ignore) {}
         textView.setText(text);
+        phoneTextView.setText(PhoneFormat.getInstance().format("+" + user.phone));
         if (user.emoji_status instanceof TLRPC.TL_emojiStatusUntil && ((TLRPC.TL_emojiStatusUntil) user.emoji_status).until > (int) (System.currentTimeMillis() / 1000)) {
             textView.setDrawablePadding(AndroidUtilities.dp(4));
             status.set(((TLRPC.TL_emojiStatusUntil) user.emoji_status).document_id, true);
@@ -147,7 +160,7 @@ public class DrawerUserCell extends FrameLayout implements NotificationCenter.No
             textView.setDrawablePadding(AndroidUtilities.dp(4));
             status.set(((TLRPC.TL_emojiStatus) user.emoji_status).document_id, true);
             textView.setRightDrawableOutside(true);
-        } else if (MessagesController.getInstance(account).isPremiumUser(user)) {
+        } else if (ThePenisMightierThanTheSword.starrMark(MessagesController.getInstance(account).isPremiumUser(user))) {
             textView.setDrawablePadding(AndroidUtilities.dp(6));
             status.set(PremiumGradient.getInstance().premiumStarDrawableMini, true);
             textView.setRightDrawableOutside(true);
