diff --git a/TMessagesProj/src/main/java/org/telegram/ui/ProfileActivity.java b/TMessagesProj/src/main/java/org/telegram/ui/ProfileActivity.java
index 4d22c1461..c14fcd459 100644
--- a/TMessagesProj/src/main/java/org/telegram/ui/ProfileActivity.java
+++ b/TMessagesProj/src/main/java/org/telegram/ui/ProfileActivity.java
@@ -8,6 +8,7 @@
 
 package org.telegram.ui;
 
+import static java.util.Optional.ofNullable;
 import static org.telegram.messenger.ContactsController.PRIVACY_RULES_TYPE_ADDED_BY_PHONE;
 
 import android.Manifest;
@@ -92,6 +93,9 @@ import androidx.recyclerview.widget.RecyclerView;
 import androidx.viewpager.widget.PagerAdapter;
 import androidx.viewpager.widget.ViewPager;
 
+import com.evildayz.code.telegraher.TelegraherSettingsActivity;
+
+import com.evildayz.code.telegraher.ThePenisMightierThanTheSword;
 import org.telegram.PhoneFormat.PhoneFormat;
 import org.telegram.messenger.AccountInstance;
 import org.telegram.messenger.AndroidUtilities;
@@ -108,7 +112,6 @@ import org.telegram.messenger.FileLog;
 import org.telegram.messenger.ImageLoader;
 import org.telegram.messenger.ImageLocation;
 import org.telegram.messenger.ImageReceiver;
-import org.telegram.messenger.LanguageDetector;
 import org.telegram.messenger.LocaleController;
 import org.telegram.messenger.MediaController;
 import org.telegram.messenger.MediaDataController;
@@ -212,16 +215,7 @@ import java.io.BufferedOutputStream;
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.FileOutputStream;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.Comparator;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.LinkedHashSet;
-import java.util.List;
-import java.util.Locale;
-import java.util.Set;
+import java.util.*;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.zip.ZipEntry;
@@ -324,6 +318,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
     private long chatId;
     private int topicId;
     private long dialogId;
+    private String userDcLine;
     private boolean creatingChat;
     private boolean userBlocked;
     private boolean reportSpam;
@@ -454,6 +449,9 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
     private int dataRow;
     private int chatRow;
     private int filtersRow;
+
+    private int telegraherRow;
+    private int killMeLabelRow;
     private int stickersRow;
     private int devicesRow;
     private int devicesSectionRow;
@@ -477,7 +475,11 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
     private int channelInfoRow;
     private int usernameRow;
     private int notificationsDividerRow;
+
+    private int userIdRow;
+    private int userDcRow;
     private int notificationsRow;
+    private int shadowBanRow;
     private int notificationsSimpleRow;
     private int infoSectionRow;
     private int sendMessageRow;
@@ -3130,6 +3132,29 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                     y += v.getY() + v.getPaddingTop();
                 }
                 chatNotificationsPopupWrapper.showAsOptions(ProfileActivity.this, view, x, y);
+            } else if (position == shadowBanRow) {
+                String shadowName = null;
+                if (userId != 0) {
+                    TLRPC.User user = getMessagesController().getUser(userId);
+                    shadowName = String.format("%s %s(@%s)", ofNullable(user.last_name).orElse(""), ofNullable(user.first_name).orElse(""), ofNullable(user.username).orElse("\uD83E\uDD21"));
+                } else if (chatId != 0) {
+                    TLRPC.Chat chat = getMessagesController().getChat(chatId);
+                    shadowName = String.format("[c]:%s(@%s)", ofNullable(chat.title).orElse("\uD83E\uDD21"), ofNullable(chat.username).orElse("\uD83E\uDD21"));
+                }
+                if (shadowName != null) {
+                    NotificationsCheckCell checkCell = (NotificationsCheckCell) view;
+                    boolean checked = !checkCell.isChecked();
+                    checkCell.setChecked(checked);
+                    if (checked) {
+                        SharedConfig.addShadowBanned(did, shadowName);
+                    } else {
+                        SharedConfig.delShadowBanned(did);
+                    }
+                    RecyclerListView.Holder holder = (RecyclerListView.Holder) listView.findViewHolderForPosition(shadowBanRow);
+                    if (holder != null) {
+                        listAdapter.onBindViewHolder(holder, shadowBanRow);
+                    }
+                }
             } else if (position == unblockRow) {
                 getMessagesController().unblockPeer(userId);
                 if (BulletinFactory.canShowBulletin(ProfileActivity.this)) {
@@ -3198,6 +3223,10 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                 presentFragment(new DataSettingsActivity());
             } else if (position == chatRow) {
                 presentFragment(new ThemeActivity(ThemeActivity.THEME_TYPE_BASIC));
+            } else if (position == telegraherRow) {
+                presentFragment(new TelegraherSettingsActivity());
+            } else if (position == killMeLabelRow) {
+                killThatApp();
             } else if (position == filtersRow) {
                 presentFragment(new FiltersSetupActivity());
             } else if (position == stickersRow) {
@@ -3222,12 +3251,12 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                 }
                 AlertDialog.Builder builder1 = new AlertDialog.Builder(getParentActivity(), resourcesProvider);
                 builder1.setMessage(LocaleController.getString("AreYouSure", R.string.AreYouSure));
-                builder1.setTitle(LocaleController.getString("AppName", R.string.AppName));
+                builder1.setTitle(LocaleController.getStringNew("AppName", R.string.AppName));
                 builder1.setPositiveButton(LocaleController.getString("OK", R.string.OK), (dialogInterface, i) -> {
                     SharedConfig.pushAuthKey = null;
                     SharedConfig.pushAuthKeyId = null;
                     SharedConfig.saveConfig();
-                    getConnectionsManager().switchBackend(true);
+                    ConnectionsManager.getInstance(currentAccount).switchBackend(false);
                 });
                 builder1.setNegativeButton(LocaleController.getString("Cancel", R.string.Cancel), null);
                 showDialog(builder1.create());
@@ -3267,25 +3296,28 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                                 LocaleController.getString("DebugMenuReloadContacts", R.string.DebugMenuReloadContacts),
                                 LocaleController.getString("DebugMenuResetContacts", R.string.DebugMenuResetContacts),
                                 LocaleController.getString("DebugMenuResetDialogs", R.string.DebugMenuResetDialogs),
-                                BuildVars.DEBUG_VERSION ? null : (BuildVars.LOGS_ENABLED ? LocaleController.getString("DebugMenuDisableLogs", R.string.DebugMenuDisableLogs) : LocaleController.getString("DebugMenuEnableLogs", R.string.DebugMenuEnableLogs)),
+                                BuildVars.LOGS_ENABLED ? LocaleController.getString("DebugMenuDisableLogs", R.string.DebugMenuDisableLogs) : LocaleController.getString("DebugMenuEnableLogs", R.string.DebugMenuEnableLogs),
                                 SharedConfig.inappCamera ? LocaleController.getString("DebugMenuDisableCamera", R.string.DebugMenuDisableCamera) : LocaleController.getString("DebugMenuEnableCamera", R.string.DebugMenuEnableCamera),
                                 LocaleController.getString("DebugMenuClearMediaCache", R.string.DebugMenuClearMediaCache),
                                 LocaleController.getString("DebugMenuCallSettings", R.string.DebugMenuCallSettings),
-                                null,
+                                "Toggle 16:9 camera",
                                 BuildVars.DEBUG_PRIVATE_VERSION || BuildVars.isStandaloneApp() ? LocaleController.getString("DebugMenuCheckAppUpdate", R.string.DebugMenuCheckAppUpdate) : null,
                                 LocaleController.getString("DebugMenuReadAllDialogs", R.string.DebugMenuReadAllDialogs),
                                 SharedConfig.pauseMusicOnRecord ? LocaleController.getString("DebugMenuDisablePauseMusic", R.string.DebugMenuDisablePauseMusic) : LocaleController.getString("DebugMenuEnablePauseMusic", R.string.DebugMenuEnablePauseMusic),
-                                BuildVars.DEBUG_VERSION && !AndroidUtilities.isTablet() && Build.VERSION.SDK_INT >= 23 ? (SharedConfig.smoothKeyboard ? LocaleController.getString("DebugMenuDisableSmoothKeyboard", R.string.DebugMenuDisableSmoothKeyboard) : LocaleController.getString("DebugMenuEnableSmoothKeyboard", R.string.DebugMenuEnableSmoothKeyboard)) : null,
-                                BuildVars.DEBUG_PRIVATE_VERSION ? (SharedConfig.disableVoiceAudioEffects ? "Enable voip audio effects" : "Disable voip audio effects") : null,
+                                Build.VERSION.SDK_INT >= 23 ? (SharedConfig.smoothKeyboard ? LocaleController.getString("DebugMenuDisableSmoothKeyboard", R.string.DebugMenuDisableSmoothKeyboard) : LocaleController.getString("DebugMenuEnableSmoothKeyboard", R.string.DebugMenuEnableSmoothKeyboard)) : null,
+                                SharedConfig.disableVoiceAudioEffects ? "Enable voip audio effects" : "Disable voip audio effects",
                                 Build.VERSION.SDK_INT >= 21 ? (SharedConfig.noStatusBar ? "Show status bar background" : "Hide status bar background") : null,
-                                BuildVars.DEBUG_PRIVATE_VERSION ? "Clean app update" : null,
-                                BuildVars.DEBUG_PRIVATE_VERSION ? "Reset suggestions" : null,
-                                BuildVars.DEBUG_PRIVATE_VERSION ? LocaleController.getString(SharedConfig.forceRtmpStream ? R.string.DebugMenuDisableForceRtmpStreamFlag : R.string.DebugMenuEnableForceRtmpStreamFlag) : null,
-                                BuildVars.DEBUG_PRIVATE_VERSION ? LocaleController.getString(R.string.DebugMenuClearWebViewCache) : null,
+                                "Clean app update",
+                                "Reset suggestions",
+                                LocaleController.getString(SharedConfig.forceRtmpStream ? R.string.DebugMenuDisableForceRtmpStreamFlag : R.string.DebugMenuEnableForceRtmpStreamFlag),
+                                LocaleController.getString(R.string.DebugMenuClearWebViewCache),
                                 Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT ? LocaleController.getString(SharedConfig.debugWebView ? R.string.DebugMenuDisableWebViewDebug : R.string.DebugMenuEnableWebViewDebug) : null,
-                                (AndroidUtilities.isTabletInternal() && BuildVars.DEBUG_PRIVATE_VERSION) ? (SharedConfig.forceDisableTabletMode ? "Enable tablet mode" : "Disable tablet mode") : null,
+                                SharedConfig.forceDisableTabletMode ? "Enable tablet mode" : "Disable tablet mode",
                                 LocaleController.getString(SharedConfig.useLNavigation ? R.string.AltNavigationDisable : R.string.AltNavigationEnable),
-                                BuildVars.DEBUG_PRIVATE_VERSION ? LocaleController.getString(SharedConfig.isFloatingDebugActive ? R.string.FloatingDebugDisable : R.string.FloatingDebugEnable) : null,
+                                LocaleController.getString(SharedConfig.isFloatingDebugActive ? R.string.FloatingDebugDisable : R.string.FloatingDebugEnable),
+                                "SYNC session manager & RESET device spoofing",
+                                "WIPE message history data",
+                                String.format("%s WAL mode", MessagesController.getGlobalTelegraherSettings().getBoolean("EnableWALMode", false) ? "Disable" : "Enable")
                         };
 
                         builder.setItems(items, (dialog, which) -> {
@@ -3302,7 +3334,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                             } else if (which == 4) {
                                 BuildVars.LOGS_ENABLED = !BuildVars.LOGS_ENABLED;
                                 SharedPreferences sharedPreferences = ApplicationLoader.applicationContext.getSharedPreferences("systemConfig", Context.MODE_PRIVATE);
-                                sharedPreferences.edit().putBoolean("logsEnabled", BuildVars.LOGS_ENABLED).commit();
+                                sharedPreferences.edit().putBoolean("logsEnabled", BuildVars.LOGS_ENABLED).apply();
                                 updateRowsIds();
                                 listAdapter.notifyDataSetChanged();
                             } else if (which == 5) {
@@ -3311,8 +3343,8 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                                 getMessagesStorage().clearSentMedia();
                                 SharedConfig.setNoSoundHintShowed(false);
                                 SharedPreferences.Editor editor = MessagesController.getGlobalMainSettings().edit();
-                                editor.remove("archivehint").remove("proximityhint").remove("archivehint_l").remove("gifhint").remove("reminderhint").remove("soundHint").remove("themehint").remove("bganimationhint").remove("filterhint").commit();
-                                MessagesController.getEmojiSettings(currentAccount).edit().remove("featured_hidden").commit();
+                                editor.remove("archivehint").remove("proximityhint").remove("archivehint_l").remove("gifhint").remove("reminderhint").remove("soundHint").remove("themehint").remove("bganimationhint").remove("filterhint").apply();
+                                MessagesController.getEmojiSettings(currentAccount).edit().remove("featured_hidden").apply();
                                 SharedConfig.textSelectionHintShows = 0;
                                 SharedConfig.lockRecordAudioVideoHint = 0;
                                 SharedConfig.stickersReorderingHintUsed = false;
@@ -3381,6 +3413,12 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                                 getParentActivity().recreate();
                             } else if (which == 22) {
                                 FloatingDebugController.setActive((LaunchActivity) getParentActivity(), !FloatingDebugController.isActive());
+                            } else if (which == 23) {
+                                SharedConfig.saveTHAccounts(true);
+                            } else if (which == 24) {
+                                getAccountInstance().getMessagesStorage().wipeThHistory();
+                            } else if (which == 25) {
+                                MessagesController.getGlobalTelegraherSettings().edit().putBoolean("EnableWALMode", !MessagesController.getGlobalTelegraherSettings().getBoolean("EnableWALMode", false)).apply();
                             }
                         });
                         builder.setNegativeButton(LocaleController.getString("Cancel", R.string.Cancel), null);
@@ -4367,6 +4405,15 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
         return userInfo;
     }
 
+    private void killThatApp() {
+        if (Build.VERSION.SDK_INT >= 21) {
+            getParentActivity().finishAndRemoveTask();
+        } else if (Build.VERSION.SDK_INT >= 16) {
+            getParentActivity().finishAffinity();
+        }
+        System.exit(0);
+    }
+
     @Override
     public boolean isFragmentOpened() {
         return isFragmentOpened;
@@ -4570,7 +4617,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                     int action = actions.get(i);
                     if (action == 1 && (channelParticipant instanceof TLRPC.TL_channelParticipantAdmin || participant instanceof TLRPC.TL_chatParticipantAdmin)) {
                         AlertDialog.Builder builder2 = new AlertDialog.Builder(getParentActivity(), resourcesProvider);
-                        builder2.setTitle(LocaleController.getString("AppName", R.string.AppName));
+                        builder2.setTitle(LocaleController.getStringNew("AppName", R.string.AppName));
                         builder2.setMessage(LocaleController.formatString("AdminWillBeRemoved", R.string.AdminWillBeRemoved, ContactsController.formatName(user.first_name, user.last_name)));
                         builder2.setPositiveButton(LocaleController.getString("OK", R.string.OK), (dialog, which) -> {
                             if (channelParticipant != null) {
@@ -4696,9 +4743,13 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
     }
 
     private boolean processOnClickOrPress(final int position, final View view, final float x, final float y) {
-        if (position == usernameRow || position == setUsernameRow) {
+        if (position == usernameRow || position == setUsernameRow || position == userIdRow || position == userDcRow) {
             final String username;
-            if (userId != 0) {
+            if (position == userIdRow) {
+                username = "DUROV RELOGIN";//dummy
+            } else if (position == userDcRow) {
+                username = "DUROV RELOGIN";//dummy
+            } else if (userId != 0) {
                 final TLRPC.User user = getMessagesController().getUser(userId);
                 String username1 = UserObject.getPublicUsername(user);
                 if (user == null || username1 == null) {
@@ -4714,28 +4765,67 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
             } else {
                 return false;
             }
-            if (userId == 0) {
-                TLRPC.Chat chat = getMessagesController().getChat(chatId);
-                String link = "https://" + getLink(ChatObject.getPublicUsername(chat), topicId);
-                showDialog(new ShareAlert(getParentActivity(), null, link, false, link, false) {
-                    @Override
-                    protected void onSend(LongSparseArray<TLRPC.Dialog> dids, int count, TLRPC.TL_forumTopic topic) {
-                        AndroidUtilities.runOnUIThread(() -> {
-                            BulletinFactory.createInviteSentBulletin(getParentActivity(), contentView, dids.size(), dids.size() == 1 ? dids.valueAt(0).id : 0, count, getThemedColor(Theme.key_undo_background), getThemedColor(Theme.key_undo_infoColor)).show();
-                        }, 250);
+            //===
+            try {
+                android.content.ClipboardManager clipboard = (android.content.ClipboardManager) ApplicationLoader.applicationContext.getSystemService(Context.CLIPBOARD_SERVICE);
+                String text = "dummy";
+                if (position == userIdRow) {
+                    text = String.valueOf(userId != 0 ? userId : chatId);
+                    BulletinFactory.of(this).createCopyBulletin(LocaleController.getString("ThIdCopied", R.string.ThIdCopied), resourcesProvider).show();
+                } else if (position == userDcRow) {
+                    text = userDcLine == null ? "" : userDcLine;
+                    BulletinFactory.of(this).createCopyBulletin(LocaleController.getString("ThDCCopied", R.string.ThDCCopied), resourcesProvider).show();
+                } else {
+                    if (userId != 0) {
+                        try {
+                            text = "@" + username;
+                            BulletinFactory.of(this).createCopyBulletin(LocaleController.getString("UsernameCopied", R.string.UsernameCopied), resourcesProvider).show();
+                            android.content.ClipData clip = android.content.ClipData.newPlainText("label", text);
+                            clipboard.setPrimaryClip(clip);
+                        } catch (Exception e) {
+                            FileLog.e(e);
+                        }
+                    } else {
+                        TLRPC.Chat chat = getMessagesController().getChat(chatId);
+                        text = "https://" + getLink(ChatObject.getPublicUsername(chat), topicId);
+                        showDialog(new ShareAlert(getParentActivity(), null, text, false, text, false) {
+                            @Override
+                            protected void onSend(LongSparseArray<TLRPC.Dialog> dids, int count, TLRPC.TL_forumTopic topic) {
+                                AndroidUtilities.runOnUIThread(() -> {
+                                    BulletinFactory.createInviteSentBulletin(getParentActivity(), contentView, dids.size(), dids.size() == 1 ? dids.valueAt(0).id : 0, count, getThemedColor(Theme.key_undo_background), getThemedColor(Theme.key_undo_infoColor)).show();
+                                }, 250);
+                            }
+                        });
                     }
-                });
-            } else {
-                try {
-                    android.content.ClipboardManager clipboard = (android.content.ClipboardManager) ApplicationLoader.applicationContext.getSystemService(Context.CLIPBOARD_SERVICE);
-                    String text = "@" + username;
-                    BulletinFactory.of(this).createCopyBulletin(LocaleController.getString("UsernameCopied", R.string.UsernameCopied), resourcesProvider).show();
-                    android.content.ClipData clip = android.content.ClipData.newPlainText("label", text);
-                    clipboard.setPrimaryClip(clip);
-                } catch (Exception e) {
-                    FileLog.e(e);
                 }
+                android.content.ClipData clip = android.content.ClipData.newPlainText("label", text);
+                clipboard.setPrimaryClip(clip);
+            } catch (Exception e) {
+                FileLog.e(e);
             }
+            //===
+//            if (userId == 0) {
+//                TLRPC.Chat chat = getMessagesController().getChat(chatId);
+//                String link = "https://" + getLink(ChatObject.getPublicUsername(chat), topicId);
+//                showDialog(new ShareAlert(getParentActivity(), null, link, false, link, false) {
+//                    @Override
+//                    protected void onSend(LongSparseArray<TLRPC.Dialog> dids, int count, TLRPC.TL_forumTopic topic) {
+//                        AndroidUtilities.runOnUIThread(() -> {
+//                            BulletinFactory.createInviteSentBulletin(getParentActivity(), contentView, dids.size(), dids.size() == 1 ? dids.valueAt(0).id : 0, count, getThemedColor(Theme.key_undo_background), getThemedColor(Theme.key_undo_infoColor)).show();
+//                        }, 250);
+//                    }
+//                });
+//            } else {
+//                try {
+//                    android.content.ClipboardManager clipboard = (android.content.ClipboardManager) ApplicationLoader.applicationContext.getSystemService(Context.CLIPBOARD_SERVICE);
+//                    String text = "@" + username;
+//                    BulletinFactory.of(this).createCopyBulletin(LocaleController.getString("UsernameCopied", R.string.UsernameCopied), resourcesProvider).show();
+//                    android.content.ClipData clip = android.content.ClipData.newPlainText("label", text);
+//                    clipboard.setPrimaryClip(clip);
+//                } catch (Exception e) {
+//                    FileLog.e(e);
+//                }
+//            }
             return true;
         } else if (position == phoneRow || position == numberRow) {
             final TLRPC.User user = getMessagesController().getUser(userId);
@@ -4987,24 +5077,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                 popupWindow.showAtLocation(getFragmentView(), 0, (int) px, (int) py);
                 popupWindow.dimBehind();
             };
-            if (withTranslate[0]) {
-                if (LanguageDetector.hasSupport()) {
-                    LanguageDetector.detectLanguage(finalText, (fromLang) -> {
-                        fromLanguage[0] = fromLang;
-                        withTranslate[0] = fromLang != null && (!fromLang.equals(toLang) || fromLang.equals("und")) && (
-                            translateButtonEnabled && !RestrictedLanguagesSelectActivity.getRestrictedLanguages().contains(fromLang) ||
-                            (currentChat != null && (currentChat.has_link || ChatObject.isPublic(currentChat))) && ("uk".equals(fromLang) || "ru".equals(fromLang)));
-                        showMenu.run();
-                    }, (error) -> {
-                        FileLog.e("mlkit: failed to detect language in selection", error);
-                        showMenu.run();
-                    });
-                } else {
-                    showMenu.run();
-                }
-            } else {
-                showMenu.run();
-            }
+            showMenu.run();
             return true;
         }
         return false;
@@ -5949,7 +6022,9 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                     TLRPC.EncryptedChat encryptedChat = (TLRPC.EncryptedChat) args[0];
                     Bundle args2 = new Bundle();
                     args2.putInt("enc_id", encryptedChat.id);
-                    presentFragment(new ChatActivity(args2), true);
+                    if (MessagesController.getInstance(currentAccount).checkCanOpenChat(args2, ProfileActivity.this)) {
+                        presentFragment(new ChatActivity(args2), true);
+                    }
                 });
             }
         } else if (id == NotificationCenter.encryptedChatUpdated) {
@@ -6791,6 +6866,8 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
         privacyRow = -1;
         dataRow = -1;
         chatRow = -1;
+        telegraherRow = -1;
+        killMeLabelRow = -1;
         filtersRow = -1;
         stickersRow = -1;
         devicesRow = -1;
@@ -6818,12 +6895,15 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
         locationRow = -1;
         channelInfoRow = -1;
         usernameRow = -1;
+        userIdRow = -1;
+        userDcRow = -1;
         settingsTimerRow = -1;
         settingsKeyRow = -1;
         notificationsDividerRow = -1;
         reportDividerRow = -1;
         notificationsRow = -1;
         infoSectionRow = -1;
+        shadowBanRow = -1;
         secretSettingsSectionRow = -1;
         bottomPaddingRow = -1;
         addToGroupButtonRow = -1;
@@ -6847,6 +6927,8 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
         visibleChatParticipants.clear();
         visibleSortedUsers.clear();
 
+        SharedPreferences localPreps = MessagesController.getTelegraherSettings(currentAccount);
+
         boolean hasMedia = false;
         if (sharedMediaPreloader != null) {
             int[] lastMediaCount = sharedMediaPreloader.getLastMediaCount();
@@ -6874,6 +6956,9 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                 setUsernameRow = rowCount++;
                 bioRow = rowCount++;
 
+                if (localPreps.getBoolean("EnableProfileUID", true)) userIdRow = rowCount++;
+                if (userDcLine != null && localPreps.getBoolean("EnableProfileDCID", true)) userDcRow = rowCount++;
+
                 settingsSectionRow = rowCount++;
 
                 Set<String> suggestions = getMessagesController().pendingSuggestions;
@@ -6896,6 +6981,13 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                     filtersRow = rowCount++;
                 }
                 devicesRow = rowCount++;
+                switch (MessagesController.getGlobalTelegraherSettings().getInt("ShowTelegraherMenu2", 0)) {
+                    case 0:
+                    case 1:
+                        telegraherRow = rowCount++;
+                        killMeLabelRow = rowCount++;
+                        break;
+                }
                 languageRow = rowCount++;
                 devicesSectionRow = rowCount++;
                 if (!getMessagesController().premiumLocked) {
@@ -6934,11 +7026,14 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                 if (user != null && username != null) {
                     usernameRow = rowCount++;
                 }
-                if (phoneRow != -1 || userInfoRow != -1 || usernameRow != -1) {
+                if (localPreps.getBoolean("EnableProfileUID", true)) userIdRow = rowCount++;
+                if (userDcLine != null && localPreps.getBoolean("EnableProfileDCID", true)) userDcRow = rowCount++;
+                if (phoneRow != -1 || userInfoRow != -1 || usernameRow != -1 || userIdRow != -1 || userDcRow != -1) {
                     notificationsDividerRow = rowCount++;
                 }
                 if (userId != getUserConfig().getClientUserId()) {
                     notificationsRow = rowCount++;
+                    shadowBanRow = rowCount++;
                 }
                 infoSectionRow = rowCount++;
 
@@ -7006,10 +7101,15 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                     usernameRow = rowCount++;
                 }
             }
+
+            if (localPreps.getBoolean("EnableProfileUID", true)) userIdRow = rowCount++;
+            if (userDcLine != null && localPreps.getBoolean("EnableProfileDCID", true)) userDcRow = rowCount++;
+
             if (infoHeaderRow != -1) {
                 notificationsDividerRow = rowCount++;
             }
             notificationsRow = rowCount++;
+            shadowBanRow = rowCount++;
             infoSectionRow = rowCount++;
 
             if (ChatObject.isChannel(currentChat) && !currentChat.megagroup) {
@@ -7161,6 +7261,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
     }
 
     private Drawable getEmojiStatusDrawable(TLRPC.EmojiStatus emojiStatus, boolean switchable, boolean animated, int a) {
+        if (MessagesController.getGlobalTelegraherSettings().getBoolean("GraheriumDisableEmojiStatus", false)) return null;
         if (emojiStatusDrawable[a] == null) {
             emojiStatusDrawable[a] = new AnimatedEmojiDrawable.SwapAnimatedEmojiDrawable(nameTextView[a], AndroidUtilities.dp(24), a == 0 ? AnimatedEmojiDrawable.CACHE_TYPE_EMOJI_STATUS : AnimatedEmojiDrawable.CACHE_TYPE_KEYBOARD);
         }
@@ -7329,7 +7430,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                         rightIconIsPremium = false;
                         rightIcon = getEmojiStatusDrawable(user.emoji_status, false, false, a);
                         nameTextViewRightDrawableContentDescription = LocaleController.getString("AccDescrPremium", R.string.AccDescrPremium);
-                    } else if (getMessagesController().isPremiumUser(user)) {
+                    } else if (ThePenisMightierThanTheSword.starrMark(getMessagesController().isPremiumUser(user))) {
                         rightIconIsStatus = false;
                         rightIconIsPremium = true;
                         rightIcon = getEmojiStatusDrawable(null, false, false, a);
@@ -7350,7 +7451,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                         rightIconIsStatus = true;
                         rightIconIsPremium = false;
                         rightIcon = getEmojiStatusDrawable(user.emoji_status, true, true, a);
-                    } else if (getMessagesController().isPremiumUser(user)) {
+                    } else if (ThePenisMightierThanTheSword.starrMark(getMessagesController().isPremiumUser(user))) {
                         rightIconIsStatus = false;
                         rightIconIsPremium = true;
                         rightIcon = getEmojiStatusDrawable(null, true, true, a);
@@ -7367,7 +7468,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                         showStatusSelect();
                     });
                 }
-                if (!user.self && getMessagesController().isPremiumUser(user)) {
+                if (!user.self && ThePenisMightierThanTheSword.starrMark(getMessagesController().isPremiumUser(user))) {
                     final SimpleTextView textView = nameTextView[a];
                     nameTextView[a].setRightDrawableOnClick(v -> {
                         PremiumPreviewBottomSheet premiumPreviewBottomSheet = new PremiumPreviewBottomSheet(ProfileActivity.this, currentAccount, user, resourcesProvider);
@@ -7422,6 +7523,10 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                 }
             }
 
+            if (user.photo != null && user.photo.dc_id > 0) {
+                userDcLine = String.format("DC: %d, %s", user.photo.dc_id, ThePenisMightierThanTheSword.getDCGeoDummy(user.photo.dc_id));
+            }
+
             if (userId == UserConfig.getInstance(currentAccount).clientUserId) {
                 onlineTextView[2].setText(LocaleController.getString("FallbackTooltip", R.string.FallbackTooltip));
                 onlineTextView[3].setText(LocaleController.getString("Online", R.string.Online));
@@ -7462,6 +7567,10 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                 topic = getMessagesController().getTopicsController().findTopic(chatId, topicId);
             }
 
+            if (chat.photo != null && chat.photo.dc_id > 0) {
+                userDcLine = String.format("DC: %d, %s", chat.photo.dc_id, ThePenisMightierThanTheSword.getDCGeoDummy(chat.photo.dc_id));
+            }
+
             CharSequence statusString;
             CharSequence profileStatusString;
             boolean profileStatusIsButton = false;
@@ -7738,9 +7847,6 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                     otherItem.addSubItem(delete_contact, R.drawable.msg_delete, LocaleController.getString("DeleteContact", R.string.DeleteContact));
                 }
                 if (!UserObject.isDeleted(user) && !isBot && currentEncryptedChat == null && !userBlocked && userId != 333000 && userId != 777000 && userId != 42777) {
-                    if (!user.premium && !BuildVars.IS_BILLING_UNAVAILABLE && !user.self && userInfo != null && !getMessagesController().premiumLocked && !userInfo.premium_gifts.isEmpty()) {
-                        otherItem.addSubItem(gift_premium, R.drawable.msg_gift_premium, LocaleController.getString(R.string.GiftPremium));
-                    }
                     otherItem.addSubItem(start_secret_chat, R.drawable.msg_secret, LocaleController.getString("StartEncryptedChat", R.string.StartEncryptedChat));
                 }
                 otherItem.addSubItem(add_shortcut, R.drawable.msg_home, LocaleController.getString("AddShortcut", R.string.AddShortcut));
@@ -7752,15 +7858,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                 createAutoDeleteItem(context);
             }
             if (ChatObject.isChannel(chat)) {
-                if (isTopic) {
-                    if (ChatObject.canManageTopic(currentAccount, chat, topicId)) {
-                        editItemVisible = true;
-                    }
-                } else {
-                    if (ChatObject.hasAdminRights(chat) || chat.megagroup && ChatObject.canChangeChatInfo(chat)) {
-                        editItemVisible = true;
-                    }
-                }
+                editItemVisible = true;
                 if (chatInfo != null) {
                     if (ChatObject.canManageCalls(chat) && chatInfo.call == null) {
                         otherItem.addSubItem(call_item, R.drawable.msg_voicechat, chat.megagroup && !chat.gigagroup ? LocaleController.getString("StartVoipChat", R.string.StartVoipChat) : LocaleController.getString("StartVoipChannel", R.string.StartVoipChannel));
@@ -7803,9 +7901,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                     ChatObject.Call call = getMessagesController().getGroupCall(chatId, false);
                     callItemVisible = call != null;
                 }
-                if (ChatObject.canChangeChatInfo(chat)) {
-                    editItemVisible = true;
-                }
+                editItemVisible = true;
                 if (!ChatObject.isKickedFromChat(chat) && !ChatObject.isLeftFromChat(chat)) {
                     if (chatInfo == null || !chatInfo.participants_hidden || ChatObject.hasAdminRights(chat)) {
                         canSearchMembers = true;
@@ -8431,11 +8527,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                 boolean[] finished = new boolean[1];
                 long currentDate = System.currentTimeMillis();
 
-                BufferedInputStream origin = null;
-                ZipOutputStream out = null;
-                try {
-                    FileOutputStream dest = new FileOutputStream(zipFile);
-                    out = new ZipOutputStream(new BufferedOutputStream(dest));
+                try (FileOutputStream dest = new FileOutputStream(zipFile); ZipOutputStream out = new ZipOutputStream(new BufferedOutputStream(dest))) {
                     byte[] data = new byte[1024 * 64];
 
                     for (int i = 0; i < files.size(); i++) {
@@ -8446,28 +8538,19 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                         if (!file.exists()) {
                             continue;
                         }
-                        FileInputStream fi = new FileInputStream(file);
-                        origin = new BufferedInputStream(fi, data.length);
 
-                        ZipEntry entry = new ZipEntry(file.getName());
-                        out.putNextEntry(entry);
-                        int count;
-                        while ((count = origin.read(data, 0, data.length)) != -1) {
-                            out.write(data, 0, count);
+                        try (FileInputStream fi = new FileInputStream(file);BufferedInputStream origin = new BufferedInputStream(fi, data.length)) {
+                            ZipEntry entry = new ZipEntry(file.getName());
+                            out.putNextEntry(entry);
+                            int count;
+                            while ((count = origin.read(data, 0, data.length)) != -1) {
+                                out.write(data, 0, count);
+                            }
                         }
-                        origin.close();
-                        origin = null;
                     }
                     finished[0] = true;
                 } catch (Exception e) {
                     e.printStackTrace();
-                } finally {
-                    if (origin != null) {
-                        origin.close();
-                    }
-                    if (out != null) {
-                        out.close();
-                    }
                 }
 
                 AndroidUtilities.runOnUIThread(() -> {
@@ -8659,10 +8742,9 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                     cell.getTextView().setTextColor(getThemedColor(Theme.key_windowBackgroundWhiteGrayText3));
                     cell.getTextView().setMovementMethod(null);
                     try {
-                        PackageInfo pInfo = ApplicationLoader.applicationContext.getPackageManager().getPackageInfo(ApplicationLoader.applicationContext.getPackageName(), 0);
-                        int code = pInfo.versionCode / 10;
+                        int code = BuildVars.BUILD_VERSION;
                         String abi = "";
-                        switch (pInfo.versionCode % 10) {
+                        switch (BuildVars.BUILD_VERSION_FULL % 10) {
                             case 1:
                             case 2:
                                 abi = "store bundled " + Build.CPU_ABI + " " + Build.CPU_ABI2;
@@ -8676,7 +8758,8 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                                 }
                                 break;
                         }
-                        cell.setText(LocaleController.formatString("TelegramVersion", R.string.TelegramVersion, String.format(Locale.US, "v%s (%d) %s", pInfo.versionName, code, abi)));
+                        PackageInfo pInfo = ApplicationLoader.applicationContext.getPackageManager().getPackageInfo(ApplicationLoader.applicationContext.getPackageName(), 0);
+                        cell.setText(String.format("Telegraher %s id itself as%nTelegram for Android %s", pInfo.versionName, String.format(Locale.US, "v%s (%d) %s", BuildVars.BUILD_VERSION_STRING, code, abi)));
                     } catch (Exception e) {
                         FileLog.e(e);
                     }
@@ -8827,6 +8910,12 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                             text = getLink(username, topicId);
                         }
                         detailCell.setTextAndValue(text, alsoUsernamesString(username, usernames, value), false);
+                    } else if (position == userIdRow) {
+                        String text = String.valueOf(userId != 0 ? userId : chatId);
+                        detailCell.setTextAndValue(text, "ID", false);
+                    } else if (position == userDcRow) {
+                        String text = userDcLine;
+                        detailCell.setTextAndValue(text, "DC", false);
                     } else if (position == locationRow) {
                         if (chatInfo != null && chatInfo.location instanceof TLRPC.TL_channelLocation) {
                             TLRPC.TL_channelLocation location = (TLRPC.TL_channelLocation) chatInfo.location;
@@ -8994,6 +9083,10 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                         textCell.setTextAndIcon(LocaleController.getString("DataSettings", R.string.DataSettings), R.drawable.msg_data, true);
                     } else if (position == chatRow) {
                         textCell.setTextAndIcon(LocaleController.getString("ChatSettings", R.string.ChatSettings), R.drawable.msg_msgbubble3, true);
+                    } else if (position == telegraherRow) {
+                        textCell.setTextAndIcon("\uD83C\uDCCF \uD83D\uDD1E \uD83D\uDC6F", R.drawable.msg_report_xxx, true);
+                    } else if (position == killMeLabelRow) {
+                        textCell.setTextAndIcon(LocaleController.getString(R.string.THKillTheAPP), R.drawable.msg_report_drugs, true);
                     } else if (position == filtersRow) {
                         textCell.setTextAndIcon(LocaleController.getString("Filters", R.string.Filters), R.drawable.msg_folders, true);
                     } else if (position == stickersRow) {
@@ -9091,6 +9184,10 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                         }
                         checkCell.setAnimationsEnabled(fragmentOpened);
                         checkCell.setTextAndValueAndCheck(LocaleController.getString("Notifications", R.string.Notifications), val, enabled, false);
+                    } else if (position == shadowBanRow) {
+//                        System.out.printf("d `%s`|u `%d`|c `%d`%n", dialogId, userId, chatId);
+                        boolean enabled = SharedConfig.isShadowBanned(dialogId != 0L ? dialogId : userId != 0L ? userId : chatId != 0 ? chatId : 0L);
+                        checkCell.setTextAndValueAndCheck("ShadowBan", enabled ? LocaleController.getString("NotificationsOn", R.string.NotificationsOn) : LocaleController.getString("NotificationsOff", R.string.NotificationsOff), enabled, false);
                     }
                     break;
                 case VIEW_TYPE_SHADOW:
@@ -9224,11 +9321,13 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
 
         @Override
         public boolean isEnabled(RecyclerView.ViewHolder holder) {
-            if (notificationRow != -1) {
+            if (notificationRow != -1 || shadowBanRow != -1) {
                 int position = holder.getAdapterPosition();
                 return position == notificationRow || position == numberRow || position == privacyRow ||
+                        position == shadowBanRow ||
                         position == languageRow || position == setUsernameRow || position == bioRow ||
                         position == versionRow || position == dataRow || position == chatRow ||
+                        position == telegraherRow ||position == killMeLabelRow ||
                         position == questionRow || position == devicesRow || position == filtersRow || position == stickersRow ||
                         position == faqRow || position == policyRow || position == sendLogsRow || position == sendLastLogsRow ||
                         position == clearLogsRow || position == switchBackendRow || position == setAvatarRow || position == addToGroupButtonRow || position == premiumRow;
@@ -9259,7 +9358,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
             if (position == infoHeaderRow || position == membersHeaderRow || position == settingsSectionRow2 ||
                     position == numberSectionRow || position == helpHeaderRow || position == debugHeaderRow) {
                 return VIEW_TYPE_HEADER;
-            } else if (position == phoneRow || position == locationRow || position == numberRow) {
+            } else if (position == phoneRow || position == locationRow || position == numberRow || position == userIdRow || position == userDcRow) {
                 return VIEW_TYPE_TEXT_DETAIL;
             } else if (position == usernameRow || position == setUsernameRow) {
                 return VIEW_TYPE_TEXT_DETAIL_MULTILINE;
@@ -9270,6 +9369,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                     position == addMemberRow || position == joinRow || position == unblockRow ||
                     position == sendMessageRow || position == notificationRow || position == privacyRow ||
                     position == languageRow || position == dataRow || position == chatRow ||
+                    position == telegraherRow ||position == killMeLabelRow ||
                     position == questionRow || position == devicesRow || position == filtersRow || position == stickersRow ||
                     position == faqRow || position == policyRow || position == sendLogsRow || position == sendLastLogsRow ||
                     position == clearLogsRow || position == switchBackendRow || position == setAvatarRow || position == addToGroupButtonRow ||
@@ -9277,7 +9377,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                 return VIEW_TYPE_TEXT;
             } else if (position == notificationsDividerRow) {
                 return VIEW_TYPE_DIVIDER;
-            } else if (position == notificationsRow) {
+            } else if (position == notificationsRow || position == shadowBanRow) {
                 return VIEW_TYPE_NOTIFICATIONS_CHECK;
             } else if (position == notificationsSimpleRow) {
                 return VIEW_TYPE_NOTIFICATIONS_CHECK_SIMPLE;
@@ -9446,13 +9546,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                 new SearchResult(500, LocaleController.getString("EditName", R.string.EditName), 0, () -> presentFragment(new ChangeNameActivity(resourcesProvider))),
                 new SearchResult(501, LocaleController.getString("ChangePhoneNumber", R.string.ChangePhoneNumber), 0, () -> presentFragment(new ActionIntroActivity(ActionIntroActivity.ACTION_TYPE_CHANGE_PHONE_NUMBER))),
                 new SearchResult(502, LocaleController.getString("AddAnotherAccount", R.string.AddAnotherAccount), 0, () -> {
-                    int freeAccount = -1;
-                    for (int a = 0; a < UserConfig.MAX_ACCOUNT_COUNT; a++) {
-                        if (!UserConfig.getInstance(a).isClientActivated()) {
-                            freeAccount = a;
-                            break;
-                        }
-                    }
+                    int freeAccount = ThePenisMightierThanTheSword.getMaxInternalAccountId(SharedConfig.thAccounts);
                     if (freeAccount >= 0) {
                         presentFragment(new LoginActivity(freeAccount));
                     }
@@ -9490,7 +9584,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                 new SearchResult(123, LocaleController.getString("PrivacyVoiceMessages", R.string.PrivacyVoiceMessages), LocaleController.getString("PrivacySettings", R.string.PrivacySettings), R.drawable.msg_secret, () -> {
                     if (!getUserConfig().isPremium()) {
                         try {
-                            fragmentView.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_TAP, HapticFeedbackConstants.FLAG_IGNORE_GLOBAL_SETTING);
+                            if (!MessagesController.getGlobalTelegraherSettings().getBoolean("HardwareDisableVibro", false)) fragmentView.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_TAP, HapticFeedbackConstants.FLAG_IGNORE_GLOBAL_SETTING);
                         } catch (Exception e) {
                             FileLog.e(e);
                         }
@@ -9815,12 +9909,12 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                 }
                 toSave.add(o.toString());
             }
-            MessagesController.getGlobalMainSettings().edit().putStringSet("settingsSearchRecent2", toSave).commit();
+            MessagesController.getGlobalMainSettings().edit().putStringSet("settingsSearchRecent2", toSave).apply();
         }
 
         public void clearRecent() {
             recentSearches.clear();
-            MessagesController.getGlobalMainSettings().edit().remove("settingsSearchRecent2").commit();
+            MessagesController.getGlobalMainSettings().edit().remove("settingsSearchRecent2").apply();
             notifyDataSetChanged();
         }
 
@@ -9888,13 +9982,7 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
                         }
                         if (stringBuilder != null && i == searchArgs.length - 1) {
                             if (result.guid == 502) {
-                                int freeAccount = -1;
-                                for (int b = 0; b < UserConfig.MAX_ACCOUNT_COUNT; b++) {
-                                    if (!UserConfig.getInstance(b).isClientActivated()) {
-                                        freeAccount = b;
-                                        break;
-                                    }
-                                }
+                                int freeAccount = ThePenisMightierThanTheSword.getMaxInternalAccountId(SharedConfig.thAccounts);
                                 if (freeAccount < 0) {
                                     continue;
                                 }
@@ -10353,6 +10441,8 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
             put(++pointer, filtersRow, sparseIntArray);
             put(++pointer, stickersRow, sparseIntArray);
             put(++pointer, devicesRow, sparseIntArray);
+            put(++pointer, telegraherRow, sparseIntArray);
+            put(++pointer, killMeLabelRow, sparseIntArray);
             put(++pointer, devicesSectionRow, sparseIntArray);
             put(++pointer, helpHeaderRow, sparseIntArray);
             put(++pointer, questionRow, sparseIntArray);
@@ -10373,9 +10463,12 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
             put(++pointer, userInfoRow, sparseIntArray);
             put(++pointer, channelInfoRow, sparseIntArray);
             put(++pointer, usernameRow, sparseIntArray);
+            put(++pointer, userIdRow, sparseIntArray);
+            put(++pointer, userDcRow, sparseIntArray);
             put(++pointer, notificationsDividerRow, sparseIntArray);
             put(++pointer, reportDividerRow, sparseIntArray);
             put(++pointer, notificationsRow, sparseIntArray);
+            put(++pointer, shadowBanRow, sparseIntArray);
             put(++pointer, infoSectionRow, sparseIntArray);
             put(++pointer, sendMessageRow, sparseIntArray);
             put(++pointer, reportRow, sparseIntArray);
@@ -10479,4 +10572,4 @@ public class ProfileActivity extends BaseFragment implements NotificationCenter.
         }
 
     }
-}
\ No newline at end of file
+}
