diff --git a/TMessagesProj/src/main/java/org/telegram/messenger/SharedConfig.java b/TMessagesProj/src/main/java/org/telegram/messenger/SharedConfig.java
index 2384918ee..72d516782 100644
--- a/TMessagesProj/src/main/java/org/telegram/messenger/SharedConfig.java
+++ b/TMessagesProj/src/main/java/org/telegram/messenger/SharedConfig.java
@@ -24,7 +24,12 @@ import android.webkit.WebView;
 import androidx.annotation.IntDef;
 import androidx.core.content.pm.ShortcutManagerCompat;
 
+import com.evildayz.code.telegraher.ThePenisMightierThanTheSword;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import org.apache.commons.lang3.StringUtils;
 import org.json.JSONObject;
+import org.telegram.PhoneFormat.PhoneFormat;
 import org.telegram.tgnet.ConnectionsManager;
 import org.telegram.tgnet.SerializedData;
 import org.telegram.tgnet.TLRPC;
@@ -37,11 +42,12 @@ import java.io.UnsupportedEncodingException;
 import java.lang.annotation.Retention;
 import java.lang.annotation.RetentionPolicy;
 import java.net.URLEncoder;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Iterator;
-import java.util.Locale;
+import java.lang.reflect.Type;
+import java.util.*;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.stream.Collectors;
+
+import static java.util.Optional.ofNullable;
 
 public class SharedConfig {
     public final static int PASSCODE_TYPE_PIN = 0,
@@ -143,7 +149,7 @@ public class SharedConfig {
     public static boolean saveStreamMedia = true;
     public static boolean smoothKeyboard = true;
     public static boolean pauseMusicOnRecord = true;
-    public static boolean chatBlur = true;
+    public static boolean chatBlur = false;
     public static boolean noiseSupression;
     public static boolean noStatusBar = true;
     public static boolean forceRtmpStream;
@@ -187,6 +193,13 @@ public class SharedConfig {
     public static boolean isFloatingDebugActive;
     public static LiteMode liteMode;
 
+    public static CopyOnWriteArraySet<Integer> activeAccounts;
+    public static Map<Integer, Map<String, String>> thAccounts;
+    public static Map<Integer, Map<String, String>> thDeviceSpoofing;
+    public static int loginingAccount = -1;
+
+    public static HashMap<Long, String> shadowBannedHM;
+
     static {
         loadConfig();
     }
@@ -280,6 +293,13 @@ public class SharedConfig {
                 editor.putInt("lockRecordAudioVideoHint", lockRecordAudioVideoHint);
                 editor.putString("storageCacheDir", !TextUtils.isEmpty(storageCacheDir) ? storageCacheDir : "");
 
+                SharedPreferences preferencesTH = ApplicationLoader.applicationContext.getSharedPreferences("telegraher", Context.MODE_PRIVATE);
+                SharedPreferences.Editor editorTH = preferencesTH.edit();
+                editorTH.putString("shadowBannedHM", ThePenisMightierThanTheSword.toJson(shadowBannedHM));
+//                editorTH.putString("thAccounts", ThePenisMightierThanTheSword.toJsonThAccounts(thAccounts));
+                editorTH.apply();
+//                System.out.println("preferencesTH saved!");
+
                 if (pendingAppUpdate != null) {
                     try {
                         SerializedData data = new SerializedData(pendingAppUpdate.getObjectSize());
@@ -317,6 +337,95 @@ public class SharedConfig {
         return value;
     }
 
+    public static void saveAccounts() {
+        FileLog.e("Save accounts: " + activeAccounts, new Exception());
+        ApplicationLoader.applicationContext.getSharedPreferences("mainconfig", Activity.MODE_PRIVATE).edit()
+                .putString("active_accounts", StringUtils.join(activeAccounts, ","))
+                .apply();
+        saveTHAccounts();
+    }
+
+    public static void saveTHAccounts() {
+        saveTHAccounts(false);
+    }
+
+    public static void saveTHAccounts(boolean reset) {
+        if (thAccounts == null || thAccounts.isEmpty()) {
+            Type thmhm = new TypeToken<Map<Integer, Map<String, Object>>>() {
+            }.getType();
+            thAccounts = new Gson().fromJson(ApplicationLoader.applicationContext.getSharedPreferences("telegraher", Activity.MODE_PRIVATE).getString("thAccounts", "{}"), thmhm);
+        }
+
+        if (reset) thAccounts.clear();
+
+        Integer[] arr = activeAccounts.stream().toArray(Integer[]::new);
+        Arrays.sort(arr);
+
+        for (int a : arr) {
+            if (!thAccounts.containsKey(a)) {
+                thAccounts.put(a, new HashMap<String, String>() {{
+                    put("userId", String.valueOf(UserConfig.getInstance(a).clientUserId));
+                    put("userPhone", PhoneFormat.getInstance().format("+" + UserConfig.getInstance(a).getUserConfig().getCurrentUser().phone));
+                    put("userFName", ofNullable(UserConfig.getInstance(a).getUserConfig().getCurrentUser().first_name).orElse(""));
+                    put("userLName", ofNullable(UserConfig.getInstance(a).getUserConfig().getCurrentUser().last_name).orElse(""));
+                    put("userName", ofNullable(UserConfig.getInstance(a).getUserConfig().getCurrentUser().username).orElse("\uD83E\uDD21"));
+                }});
+            }
+        }
+
+        if (!thAccounts.containsKey(-1)) {
+            thAccounts.put(-1, new HashMap<String, String>() {{
+                put("nextAccountId", String.valueOf(ThePenisMightierThanTheSword.getMaxInternalAccountId(thAccounts) + 1));
+                put("thAccountVersion", "1");
+            }});
+        } else {
+            ThePenisMightierThanTheSword.getMaxInternalAccountId(thAccounts);
+        }
+
+        ApplicationLoader.applicationContext.getSharedPreferences("telegraher", Context.MODE_PRIVATE).edit()
+                .putString("thAccounts", ThePenisMightierThanTheSword.toJsonNestedMaps(thAccounts))
+                .apply();
+        saveTHDeviceSpoofing(reset);
+    }
+
+    public static void saveTHDeviceSpoofing() {
+        saveTHDeviceSpoofing(false);
+    }
+
+    public static void saveTHDeviceSpoofing(boolean reset) {
+        if (thDeviceSpoofing == null || thDeviceSpoofing.isEmpty()) {
+            Type thmhm = new TypeToken<Map<Integer, Map<String, String>>>() {
+            }.getType();
+            thDeviceSpoofing = new Gson().fromJson(ApplicationLoader.applicationContext.getSharedPreferences("telegraher", Activity.MODE_PRIVATE).getString("thDeviceSpoofing", "{}"), thmhm);
+        }
+
+        if (reset) thDeviceSpoofing.clear();
+
+        if (!thDeviceSpoofing.containsKey(-1)) {
+            thDeviceSpoofing.put(-1,
+                    new HashMap<String, String>() {{
+                        put("deviceBrand", Build.MANUFACTURER);
+                        put("deviceModel", Build.MODEL);
+                        put("deviceSDK", Integer.valueOf(Build.VERSION.SDK_INT).toString());
+                    }});
+        }
+
+        Set<Integer> ids;
+        if ((ids = thAccounts.keySet()).isEmpty()) ids = activeAccounts;
+
+        for (int a : ids) {
+            if (!thDeviceSpoofing.containsKey(a)) {
+                thDeviceSpoofing.put(a, new HashMap<String, String>() {{
+                    putAll(thDeviceSpoofing.get(-1));
+                }});
+            }
+        }
+
+        ApplicationLoader.applicationContext.getSharedPreferences("telegraher", Context.MODE_PRIVATE).edit()
+                .putString("thDeviceSpoofing", ThePenisMightierThanTheSword.toJsonNestedMaps(thDeviceSpoofing))
+                .apply();
+    }
+
     public static void loadConfig() {
         synchronized (sync) {
             if (configLoaded || ApplicationLoader.applicationContext == null) {
@@ -372,7 +481,7 @@ public class SharedConfig {
                         data.cleanup();
                     }
                 }
-                if (pendingAppUpdate != null) {
+                if (false) {
                     long updateTime = 0;
                     int updateVersion = 0;
                     String updateVersionString = null;
@@ -429,7 +538,7 @@ public class SharedConfig {
             saveStreamMedia = preferences.getBoolean("saveStreamMedia", true);
             smoothKeyboard = preferences.getBoolean("smoothKeyboard2", true);
             pauseMusicOnRecord = preferences.getBoolean("pauseMusicOnRecord", false);
-            chatBlur = preferences.getBoolean("chatBlur", true);
+            chatBlur = preferences.getBoolean("chatBlur", false);
             forceDisableTabletMode = preferences.getBoolean("forceDisableTabletMode", false);
             streamAllVideo = preferences.getBoolean("streamAllVideo", BuildVars.DEBUG_VERSION);
             streamMkv = preferences.getBoolean("streamMkv", false);
@@ -463,6 +572,13 @@ public class SharedConfig {
             messageSeenHintCount = preferences.getInt("messageSeenCount", 3);
             emojiInteractionsHintCount = preferences.getInt("emojiInteractionsHintCount", 3);
             dayNightThemeSwitchHintCount = preferences.getInt("dayNightThemeSwitchHintCount", 3);
+            activeAccounts = Arrays.stream(preferences.getString("active_accounts", "").split(",")).filter(StringUtils::isNotBlank).map(Integer::parseInt).collect(Collectors.toCollection(CopyOnWriteArraySet::new));
+            if (!preferences.contains("activeAccountsLoaded")) {
+                if (!SharedConfig.activeAccounts.isEmpty()) {
+                    preferences.edit().putString("active_accounts", StringUtils.join(activeAccounts, ",")).apply();
+                }
+                preferences.edit().putBoolean("activeAccountsLoaded", true).apply();
+            }
             mediaColumnsCount = preferences.getInt("mediaColumnsCount", 3);
             fastScrollHintCount = preferences.getInt("fastScrollHintCount", 3);
             dontAskManageStorage = preferences.getBoolean("dontAskManageStorage", false);
@@ -473,6 +589,13 @@ public class SharedConfig {
             preferences = ApplicationLoader.applicationContext.getSharedPreferences("Notifications", Activity.MODE_PRIVATE);
             showNotificationsForAllAccounts = preferences.getBoolean("AllAccounts", true);
 
+            preferences = ApplicationLoader.applicationContext.getSharedPreferences("telegraher", Context.MODE_PRIVATE);
+            Type lhm = new TypeToken<HashMap<Long, String>>() {}.getType();
+            shadowBannedHM = new Gson().fromJson(preferences.getString("shadowBannedHM", "{}"), lhm);
+            Type thmhm = new TypeToken<Map<Integer, Map<String, String>>>() {}.getType();
+            thAccounts=new Gson().fromJson(preferences.getString("thAccounts","{}"),thmhm);
+            thDeviceSpoofing=new Gson().fromJson(preferences.getString("thDeviceSpoofing","{}"),thmhm);
+
             configLoaded = true;
 
             try {
@@ -533,6 +656,50 @@ public class SharedConfig {
         getCountryLangs();
     }
 
+    public static HashMap<Long, String> getShadowBannedHM() {
+        if (shadowBannedHM == null) shadowBannedHM = new HashMap<>();
+        return shadowBannedHM;
+    }
+
+    public static boolean isShadowBanned(TLRPC.Message m) {
+//        System.out.printf("isShadowBanned `%s`%n", "the message");
+        if (shadowBannedHM == null) {
+            shadowBannedHM = new HashMap<>();
+            return false;
+        }
+        return isShadowBanned(m.from_id.channel_id) || isShadowBanned(m.from_id.chat_id) || isShadowBanned(m.from_id.user_id);
+    }
+
+    public static boolean isShadowBanned(TLRPC.Updates m) {
+//        System.out.printf("isShadowBanned `%s`%n", "the Updates");
+        if (shadowBannedHM == null) {
+            shadowBannedHM = new HashMap<>();
+            return false;
+        }
+        return m instanceof TLRPC.TL_updateShortChatMessage ? isShadowBanned(m.from_id) : isShadowBanned(m.user_id);
+    }
+
+    public static boolean isShadowBanned(long id) {
+//        System.out.printf("isShadowBanned `%d`%n", id);
+        if (shadowBannedHM == null) {
+            shadowBannedHM = new HashMap<>();
+            return false;
+        }
+        return shadowBannedHM.containsKey(id < 0L ? -id : id);
+    }
+
+    public static void addShadowBanned(long id, String name) {
+//        System.out.printf("addShadowBanned `%d`%n", id);
+        if (shadowBannedHM == null) shadowBannedHM = new HashMap<>();
+        shadowBannedHM.put(id < 0L ? -id : id, name);
+    }
+
+    public static void delShadowBanned(long id) {
+//        System.out.printf("delShadowBanned `%d`%n", id);
+        if (shadowBannedHM == null) shadowBannedHM = new HashMap<>();
+        shadowBannedHM.remove(id < 0L ? -id : id);
+    }
+
     public static HashMap<String, String> getCountryLangs() {
         if (passportConfigMap == null) {
             passportConfigMap = new HashMap<>();
@@ -551,6 +718,7 @@ public class SharedConfig {
     }
 
     public static boolean isAppUpdateAvailable() {
+        if (true) return false;
         if (pendingAppUpdate == null || pendingAppUpdate.document == null || !BuildVars.isStandaloneApp()) {
             return false;
         }
@@ -566,6 +734,7 @@ public class SharedConfig {
     }
 
     public static boolean setNewAppVersionAvailable(TLRPC.TL_help_appUpdate update) {
+        if (true) return false;
         String updateVersionString = null;
         int versionCode = 0;
         try {
@@ -655,7 +824,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putInt("suggestStickers", suggestStickers);
-        editor.commit();
+        editor.apply();
     }
 
     public static void setSearchMessagesAsListUsed(boolean value) {
@@ -663,7 +832,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("searchMessagesAsListUsed", searchMessagesAsListUsed);
-        editor.commit();
+        editor.apply();
     }
 
     public static void setStickersReorderingHintUsed(boolean value) {
@@ -671,28 +840,28 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("stickersReorderingHintUsed", stickersReorderingHintUsed);
-        editor.commit();
+        editor.apply();
     }
 
     public static void increaseTextSelectionHintShowed() {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putInt("textSelectionHintShows", ++textSelectionHintShows);
-        editor.commit();
+        editor.apply();
     }
 
     public static void removeTextSelectionHint() {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putInt("textSelectionHintShows", 3);
-        editor.commit();
+        editor.apply();
     }
 
     public static void increaseScheduledOrNoSuoundHintShowed() {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putInt("scheduledOrNoSoundHintShows", ++scheduledOrNoSoundHintShows);
-        editor.commit();
+        editor.apply();
     }
 
     public static void forwardingOptionsHintHintShowed() {
@@ -700,35 +869,35 @@ public class SharedConfig {
         SharedPreferences.Editor editor = preferences.edit();
         forwardingOptionsHintShown = true;
         editor.putBoolean("forwardingOptionsHintShown", forwardingOptionsHintShown);
-        editor.commit();
+        editor.apply();
     }
 
     public static void removeScheduledOrNoSoundHint() {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putInt("scheduledOrNoSoundHintShows", 3);
-        editor.commit();
+        editor.apply();
     }
 
     public static void increaseLockRecordAudioVideoHintShowed() {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putInt("lockRecordAudioVideoHint", ++lockRecordAudioVideoHint);
-        editor.commit();
+        editor.apply();
     }
 
     public static void removeLockRecordAudioVideoHint() {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putInt("lockRecordAudioVideoHint", 3);
-        editor.commit();
+        editor.apply();
     }
 
     public static void increaseSearchAsListHintShows() {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putInt("searchMessagesAsListHintShows", ++searchMessagesAsListHintShows);
-        editor.commit();
+        editor.apply();
     }
 
     public static void setKeepMedia(int value) {
@@ -736,7 +905,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putInt("keep_media", keepMedia);
-        editor.commit();
+        editor.apply();
     }
 
     public static void checkLogsToDelete() {
@@ -760,7 +929,7 @@ public class SharedConfig {
             SharedPreferences preferences = MessagesController.getGlobalMainSettings();
             SharedPreferences.Editor editor = preferences.edit();
             editor.putInt("lastLogsCheckTime", lastLogsCheckTime);
-            editor.commit();
+            editor.apply();
         });
     }
 
@@ -775,7 +944,7 @@ public class SharedConfig {
         Utilities.cacheClearQueue.postRunnable(() -> {
             boolean hasExceptions = false;
             ArrayList<CacheByChatsController> cacheByChatsControllers = new ArrayList<>();
-            for (int account = 0; account < UserConfig.MAX_ACCOUNT_COUNT; account++) {
+            for (int account : SharedConfig.activeAccounts) {
                 if (UserConfig.getInstance(account).isClientActivated()) {
                     CacheByChatsController cacheByChatsController = UserConfig.getInstance(account).getMessagesController().getCacheByChatsController();
                     cacheByChatsControllers.add(cacheByChatsController);
@@ -862,7 +1031,7 @@ public class SharedConfig {
                 if (maxCacheGb == 1) {
                     maxCacheSize = 1024L * 1024L * 300L;
                 } else {
-                    maxCacheSize = maxCacheGb * 1024L * 1024L * 1000L;
+                    maxCacheSize = maxCacheGb * 1024L * 1024L * 1024L;
                 }
                 final SparseArray<File> paths = ImageLoader.getInstance().createMediaPaths();
                 long totalSize = 0;
@@ -943,7 +1112,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("disableVoiceAudioEffects", disableVoiceAudioEffects);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleNoiseSupression() {
@@ -951,7 +1120,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("noiseSupression", noiseSupression);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleForceRTMPStream() {
@@ -978,7 +1147,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("noStatusBar", noStatusBar);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleLoopStickers() {
@@ -986,7 +1155,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("loopStickers", loopStickers);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleBigEmoji() {
@@ -994,7 +1163,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("allowBigEmoji", allowBigEmoji);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleSuggestAnimatedEmoji() {
@@ -1002,7 +1171,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("suggestAnimatedEmoji", suggestAnimatedEmoji);
-        editor.commit();
+        editor.apply();
     }
 
     public static void setPlaybackOrderType(int type) {
@@ -1021,7 +1190,7 @@ public class SharedConfig {
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("shuffleMusic", shuffleMusic);
         editor.putBoolean("playOrderReversed", playOrderReversed);
-        editor.commit();
+        editor.apply();
     }
 
     public static void setRepeatMode(int mode) {
@@ -1032,7 +1201,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putInt("repeatMode", repeatMode);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleSaveToGalleryFlag(int flag) {
@@ -1054,7 +1223,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("autoplay_gif", autoplayGifs);
-        editor.commit();
+        editor.apply();
     }
 
     public static void setUseThreeLinesLayout(boolean value) {
@@ -1062,7 +1231,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("useThreeLinesLayout", useThreeLinesLayout);
-        editor.commit();
+        editor.apply();
         NotificationCenter.getGlobalInstance().postNotificationName(NotificationCenter.dialogsNeedReload, true);
     }
 
@@ -1071,7 +1240,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("archiveHidden", archiveHidden);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleAutoplayVideo() {
@@ -1079,7 +1248,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("autoplay_video", autoplayVideo);
-        editor.commit();
+        editor.apply();
     }
 
     public static boolean isSecretMapPreviewSet() {
@@ -1092,7 +1261,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putInt("mapPreviewType", mapPreviewType);
-        editor.commit();
+        editor.apply();
     }
 
     public static void setNoSoundHintShowed(boolean value) {
@@ -1103,7 +1272,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("noSoundHintShowed", noSoundHintShowed);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toogleRaiseToSpeak() {
@@ -1111,7 +1280,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("raise_to_speak", raiseToSpeak);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleCustomTabs() {
@@ -1119,7 +1288,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("custom_tabs", customTabs);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleDirectShare() {
@@ -1127,7 +1296,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("direct_share", directShare);
-        editor.commit();
+        editor.apply();
         ShortcutManagerCompat.removeAllDynamicShortcuts(ApplicationLoader.applicationContext);
         MediaDataController.getInstance(UserConfig.selectedAccount).buildShortcuts();
     }
@@ -1137,7 +1306,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("streamMedia", streamMedia);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleSortContactsByName() {
@@ -1145,7 +1314,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("sortContactsByName", sortContactsByName);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleSortFilesByName() {
@@ -1153,7 +1322,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("sortFilesByName", sortFilesByName);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleStreamAllVideo() {
@@ -1161,7 +1330,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("streamAllVideo", streamAllVideo);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleStreamMkv() {
@@ -1169,7 +1338,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("streamMkv", streamMkv);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleSaveStreamMedia() {
@@ -1177,7 +1346,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("saveStreamMedia", saveStreamMedia);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleSmoothKeyboard() {
@@ -1185,7 +1354,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("smoothKeyboard2", smoothKeyboard);
-        editor.commit();
+        editor.apply();
     }
 
     public static void togglePauseMusicOnRecord() {
@@ -1193,7 +1362,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("pauseMusicOnRecord", pauseMusicOnRecord);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleChatBlur() {
@@ -1201,7 +1370,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("chatBlur", chatBlur);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleForceDisableTabletMode() {
@@ -1209,7 +1378,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("forceDisableTabletMode", forceDisableTabletMode);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleInappCamera() {
@@ -1217,7 +1386,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("inappCamera", inappCamera);
-        editor.commit();
+        editor.apply();
     }
 
     public static void toggleRoundCamera16to9() {
@@ -1225,7 +1394,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putBoolean("roundCamera16to9", roundCamera16to9);
-        editor.commit();
+        editor.apply();
     }
 
     public static void setDistanceSystemType(int type) {
@@ -1233,7 +1402,7 @@ public class SharedConfig {
         SharedPreferences preferences = MessagesController.getGlobalMainSettings();
         SharedPreferences.Editor editor = preferences.edit();
         editor.putInt("distanceSystemType", distanceSystemType);
-        editor.commit();
+        editor.apply();
         LocaleController.resetImperialSystemType();
     }
 
@@ -1291,7 +1460,7 @@ public class SharedConfig {
             serializedData.writeString(info.secret != null ? info.secret : "");
         }
         SharedPreferences preferences = ApplicationLoader.applicationContext.getSharedPreferences("mainconfig", Activity.MODE_PRIVATE);
-        preferences.edit().putString("proxy_list", Base64.encodeToString(serializedData.toByteArray(), Base64.NO_WRAP)).commit();
+        preferences.edit().putString("proxy_list", Base64.encodeToString(serializedData.toByteArray(), Base64.NO_WRAP)).apply();
         serializedData.cleanup();
     }
 
@@ -1322,7 +1491,7 @@ public class SharedConfig {
             editor.putInt("proxy_port", 1080);
             editor.putBoolean("proxy_enabled", false);
             editor.putBoolean("proxy_enabled_calls", false);
-            editor.commit();
+            editor.apply();
             if (enabled) {
                 ConnectionsManager.setProxySettings(false, "", 0, "", "", "");
             }
@@ -1467,7 +1636,7 @@ public class SharedConfig {
     }
 
     public static boolean canBlurChat() {
-        return getDevicePerformanceClass() == PERFORMANCE_CLASS_HIGH;
+        return true;
     }
 
     public static boolean chatBlurEnabled() {
